containers/src/Data/IntMap/Internal.hs

```diff
@@ -62,2837 +275,1076 @@
+xorBounds (Bound min) (Bound max) = Data.Bits.xor (i2w min) (i2w max)
 
--- See Note: Order of constructors
-data IntMap a = Bin {-# UNPACK #-} !Prefix
-                    {-# UNPACK #-} !Mask
-                    !(IntMap a)
-                    !(IntMap a)
--- Fields:
---   prefix: The most significant bits shared by all keys in this Bin.
---   mask: The switching bit to determine if a key should follow the left
---         or right subtree of a 'Bin'.
--- Invariant: Nil is never found as a child of Bin.
--- Invariant: The Mask is a power of 2. It is the largest bit position at which
---            two keys of the map differ.
--- Invariant: Prefix is the common high-order bits that all elements share to
---            the left of the Mask bit.
--- Invariant: In (Bin prefix mask left right), left consists of the elements that
---            don't have the mask bit set; right is all the elements that do.
-              | Tip {-# UNPACK #-} !Key a
-              | Nil
-
-type Prefix = Int
-type Mask   = Int
-
-
--- Some stuff from "Data.IntSet.Internal", for 'restrictKeys' and
--- 'withoutKeys' to use.
-type IntSetPrefix = Int
-type IntSetBitMap = Word
-
-bitmapOf :: Int -> IntSetBitMap
-bitmapOf x = shiftLL 1 (x .&. IntSet.suffixBitMask)
-{-# INLINE bitmapOf #-}
-
-{--------------------------------------------------------------------
-  Operators
---------------------------------------------------------------------}
+{-# INLINE boundsDisjoint #-}
+boundsDisjoint :: Bound L -> Bound R -> Bool
+boundsDisjoint (Bound min) (Bound max) = min > max
 
--- | /O(min(n,W))/. Find the value at a key.
--- Calls 'error' when the element can not be found.
---
--- > fromList [(5,'a'), (3,'b')] ! 1    Error: element not in the map
--- > fromList [(5,'a'), (3,'b')] ! 5 == 'a'
+-- Phantom types used to separate the types of left and right nodes.
+-- They are uninhabited simply to ensure that they are only used as type parameters.
+data L
+data R
 
-(!) :: IntMap a -> Key -> a
-(!) m k = find k m
+#if defined(__GLASGOW_HASKELL__)
+-- TODO: If we are relying on GHC features anyway, L and R could be a new kind.
+newtype Bound t = Bound { boundKey :: Key } deriving (Eq, Ord, Show)
 
--- | /O(min(n,W))/. Find the value at a key.
--- Returns 'Nothing' when the element can not be found.
---
--- > fromList [(5,'a'), (3,'b')] !? 1 == Nothing
--- > fromList [(5,'a'), (3,'b')] !? 5 == Just 'a'
---
--- @since 0.5.11
+type family Flipped t
+type instance Flipped L = R
+type instance Flipped R = L
+#else
+-- Without type families, we can't track min vs. max correctly, so we just don't by making that parameter ignored
+type Bound t = Bound_
+newtype Bound_ = Bound { boundKey :: Key } deriving (Eq, Ord, Show)
+-- This, like L and R, is uninhabited to ensure that it is only used as a type parameter
+data Flipped t
+#endif
 
-(!?) :: IntMap a -> Key -> Maybe a
-(!?) m k = lookup k m
+inMinBound :: Key -> Bound L -> Bool
+inMinBound k (Bound min) = k > min
```

Comment:
We almost always need to do something different if the key is equal to the bound instead of actually inside the bound. The typical pattern is

```haskell
function k min other parameters
    | inMinBound k min -> continue navigating
    | outOfMinBound k min -> key not found
    | otherwise -> key found
```

I agree that the "in bounds" terminology isn't completely clear, and I should probably make explicit it being exclusive, but i couldn't think of anything better.

Reply:
Ah thanks for the explanation. If we keep `inMinBound`, please include the explanation in the docs.

How about using a proper sum type for the three cases though? Something like

    data InBound = OutOfBound | ExactMatch | InBound

? Would that degrade performance?